pkg/agent/agent.go

```diff
@@ -663,6 +674,62 @@ func (c *agentController) SyncAddOnPlacementScore(ctx context.Context) error {
 	return nil
 }
 
+func (c *agentController) deleteManagedCluster(ctx context.Context, hc *hyperv1beta1.HostedCluster) error {
+	if hc == nil {
+		return fmt.Errorf("failed to delete nil hostedCluster")
+	}
+
+	managedClusterName, ok := hc.GetAnnotations()[util.ManagedClusterAnnoKey]
+	if !ok || len(managedClusterName) == 0 {
+		managedClusterName = hc.Name
+	}
+
+	klusterletName := "klusterlet-" + managedClusterName
+
+	// Remove the operator.open-cluster-management.io/klusterlet-hosted-cleanup finalizer in klusterlet
+	klusterlet := &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: klusterletName,
+		},
+	}
+
+	if err := c.spokeClient.Get(ctx, client.ObjectKeyFromObject(klusterlet), klusterlet); err != nil {
+		if apierrors.IsNotFound(err) {
+			c.log.Info(fmt.Sprintf("klusterlet %v is already deleted", klusterletName))
+		} else {
+			c.log.Error(err, fmt.Sprintf("failed to get the klusterlet %v", klusterletName))
+			return err
+		}
+	}
+
+	updated := controllerutil.RemoveFinalizer(klusterlet, "operator.open-cluster-management.io/klusterlet-hosted-cleanup")
```

Comment:
Hi @philipwu08 , it is better to remove the finalizer when the klusterlet is in deletion status, otherwise, the klusterlet-operator will add it back.

Reply:
BTW, deleting managed cluster will trigger to delete the klusterlet.  Not sure if it is needed to watch the klusterlet resources which have the `operator.open-cluster-management.io/klusterlet-hosted-cleanup` finalizer

Reply:
Making hypershift-addon-agent watch klusterlet seems like an unreasonable task. After discussing with Qiu Jian, we will enhance the klusterlet operator and add an eviction mechanism. If the cluster is unreachable for 5 minutes, we will forcefully remove the finalizer.